xoa_driver/functions/anlt.py

```diff
@@ -69,22 +85,42 @@ def obtain_ports_of_module(
     return ports
 
 
-def obtain_port(
-    tester: GenericAnyTester,
-    module_id: int,
-    port_id: int,
-) -> GenericAnyPort:
+def obtain_ports(
+    tester: GenericAnyTester, module_id: int = -1, port_id: int = -1
+) -> tuple[GenericAnyPort]:
     if tester is None:
         raise NotConnectedError()
-    try:
-        module = tester.modules.obtain(module_id)
-    except KeyError:
-        raise NoSuchModuleError(module_id)
-    try:
-        port = module.ports.obtain(port_id)
-    except KeyError:
-        raise NoSuchModuleError(port_id)
-    return port
+    if module_id == -1:
+        modules = tuple(tester.modules)
+    else:
+        try:
+            modules = tester.modules.obtain_multiple(module_id)
+        except KeyError:
+            raise NoSuchModuleError(module_id)
+    ports = []
+    for module in modules:
+        if port_id == -1:
+            ports.extend(list(module.ports))
+        else:
+            try:
+                port = module.ports.obtain(port_id)
+                ports.append(port)
+            except KeyError:
+                raise NoSuchPortError(port_id)
+
+    return tuple(ports)
```

Comment:
The logic of this function is to coupled, and must be spliced out to sub methods, or simplified.